src/Microsoft.AspNetCore.Blazor.Razor.Extensions/TrimWhitespacePass.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+using System;
+
+namespace Microsoft.AspNetCore.Blazor.Razor
+{
+    internal class TrimWhitespacePass : IntermediateNodePassBase, IRazorDirectiveClassifierPass
+    {
+        protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+        {
+            if (codeDocument == null)
+            {
+                throw new ArgumentNullException(nameof(codeDocument));
+            }
+
+            if (documentNode == null)
+            {
+                throw new ArgumentNullException(nameof(documentNode));
+            }
+
+            var method = documentNode.FindPrimaryMethod();
+            if (method != null)
+            {
+                RemoveContiguousWhitespace(method.Children, TraversalDirection.Forwards);
+                RemoveContiguousWhitespace(method.Children, TraversalDirection.Backwards);
+            }
+        }
+
+        private static void RemoveContiguousWhitespace(IntermediateNodeCollection nodes, TraversalDirection direction)
+        {
+            var position = direction == TraversalDirection.Forwards ? 0 : nodes.Count - 1;
+            while (position >= 0 && position < nodes.Count)
+            {
+                var node = nodes[position];
+                bool shouldRemoveNode;
+                bool shouldContinueIteration;
+
+                switch (node)
+                {
+                    case IntermediateToken intermediateToken:
+                        shouldRemoveNode = string.IsNullOrWhiteSpace(intermediateToken.Content);
+                        shouldContinueIteration = shouldRemoveNode;
+                        break;
+
+                    case HtmlContentIntermediateNode htmlContentIntermediateNode:
+                        RemoveContiguousWhitespace(htmlContentIntermediateNode.Children, direction);
+                        shouldRemoveNode = htmlContentIntermediateNode.Children.Count == 0;
+                        shouldContinueIteration = shouldRemoveNode;
+                        break;
+
+                    case CSharpExpressionIntermediateNode _:
+                        shouldRemoveNode = false;
+                        shouldContinueIteration = false; // Because it may produce nonwhitespace output
+                        break;
```

Comment:
Can you try something in VS with this change?

Write something like:
```
<div class=@( )>

</div>
```

Then add:
```
<div class=@( )>

</div>
@functions {
  string myclass;
}
```

The put your cursor here:
```
<div class=@(| )>

</div>
@functions {
  string myclass;
}
```

Do you get completion for `myclass` at that location?

I'm concerned that this is a case where design time will have to function differently

Reply:
Yes, I do still get completion in that case.

I'm interested to know what you're suspicious might be broken though in case I'm not understanding the rules. How would removing some whitespace-only `IntermediateNode` instances from the beginning and end of the method change completion?

Reply:
OK great.

This concern was based on my earlier belief that you wanted to remove whitespace inside C#. The problem with that at design time is that we need a span in the generated code that maps to where your cursor is. If your cursor is inside C# whitespace then we need to generate that whitespace and make sure it matches so that the right completion will trigger.

If you are observing `documentNode.Options.DesignTime` from inside VS, that will **always** be the case. The code generation of the files we use in the build is driven by msbuild, which is a separate process.

You should be able to see `DesignTime == false` in any of the tests for runtime code generation.

https://github.com/aspnet/Blazor/blob/dev/test/Microsoft.AspNetCore.Blazor.Build.Test/RazorIntegrationTestBase.cs#L201

Reply:
I think it would be fine to skip over this detail, or to make this pass no-op for design time if you think that makes sense

Reply:
OK, I've now updated it to skip `TrimWhitespacePass` both during design-time builds *and* during the first phase of the two-phase compile, since it wasn't achieving anything useful during those compilations.